### PR TITLE
added error on product creation when utl key is duplicated

### PR DESCRIPTION
**Pull request by riccardougolini**
#20434 
check out for this solution.

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @riccardougolini. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by rogyar** [comment]
Hi @riccardougolini. Thank you for your contribution. Could you provide more information about your solution, please? There's an existing PR (https://github.com/magento/magento2/pull/20597) that fixes the original issue. If your solution has some extra points, please describe it.
Thank you.

**Comment by riccardougolini** [comment]
hi @rogyar , this type of solution is applied at a low level so that the speed of the site is not affected. As you can see i applied a backend model on url key product. This backend type just check if it already exists same url-key. No observer, no preference. Just resource model and data patch upgrade.

**Comment by hostep** [comment]
@riccardougolini: just a remark, but I think you'll need to add the `store_id` to the where condition in your`IsUniqueUrlKey` class as well. A duplicated url key can exist, as long as it lives in a different store view.

File: app/code/Magento/CatalogUrlRewrite/Model/Product/Attribute/Backend/UrlKey.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        IsUniqueUrlKey $urlKey
+    ) {
+        $this->urlKey = $urlKey;
+    }
+
+    /**
+     * @param  Product $object
+     * @return bool
+     * @throws \Magento\Framework\Exception\LocalizedException
+     * @throws \Magento\Framework\Exception\LocalizedException
```
Comment (williankeller) [inline app/code/Magento/CatalogUrlRewrite/Model/Product/Attribute/Backend/UrlKey.php]: HI @riccardougolini
Duplicated `@throws` here.

File: app/code/Magento/CatalogUrlRewrite/Setup/Patch/Data/UpdateUrlKeyProductAttributePatch.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    public function __construct(
+        ModuleDataSetupInterface $moduleDataSetup,
+        CategorySetupFactory $categorySetupFactory
+    ) {
+        $this->moduleDataSetup = $moduleDataSetup;
+        $this->categorySetupFactory = $categorySetupFactory;
+    }
+
+    /**
+     * {@inheritdoc}
```
Comment (williankeller) [inline app/code/Magento/CatalogUrlRewrite/Setup/Patch/Data/UpdateUrlKeyProductAttributePatch.php]: HI @riccardougolini, `@inheritdoc` should be without brackets

File: app/code/Magento/CatalogUrlRewrite/Model/Product/Attribute/Backend/UrlKey.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
+<?php
+
```
Comment (williankeller) [inline app/code/Magento/CatalogUrlRewrite/Model/Product/Attribute/Backend/UrlKey.php]: Hi @riccardougolini 
Missing `declare(strict_types=1);`

**Review by williankeller — CHANGES_REQUESTED** [review]

File: app/code/Magento/CatalogUrlRewrite/Model/Product/Attribute/Backend/UrlKey.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        $this->urlKey = $urlKey;
+    }
+
+    /**
+     * @param  Product $object
+     * @return bool
+     * @throws \Magento\Framework\Exception\LocalizedException
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    public function validate($object)
```
Comment (williankeller) [inline app/code/Magento/CatalogUrlRewrite/Model/Product/Attribute/Backend/UrlKey.php:32]: Hi @riccardougolini,
Also missing argument typing and return typing

Comment (riccardougolini) [reply app/code/Magento/CatalogUrlRewrite/Model/Product/Attribute/Backend/UrlKey.php:32]: hi @williankeller, unfortunately I can't type the method more than this cause on \Magento\Eav\Model\Entity\Attribute\Backend\AbstractBackend the method validate does not have a complete typing.

File: app/code/Magento/CatalogUrlRewrite/Test/Integration/Model/Product/Attribute/Backend/IsUniqueUrlKeyTest.php [hunk 5]
```diff
@@ -0,0 +1,111 @@
+<?php
+
```
Comment (rogyar) [inline app/code/Magento/CatalogUrlRewrite/Test/Integration/Model/Product/Attribute/Backend/IsUniqueUrlKeyTest.php]: Please, remove extra lines, add strict_types declaration and Magento copyright

File: app/code/Magento/CatalogUrlRewrite/Test/Integration/_fixtures/product_url_key_attribute.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
+<?php
+
```
Comment (rogyar) [inline app/code/Magento/CatalogUrlRewrite/Test/Integration/_fixtures/product_url_key_attribute.php]: Please, add strict_types declaration and Magento copyright

File: app/code/Magento/CatalogUrlRewrite/Test/Integration/Model/Product/Attribute/Backend/IsUniqueUrlKeyTest.php [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+{
+    /**
+     * @var ProductRepositoryInterface $productRepository
+     */
+    private $productRepository;
+
+    /**
+     * @var CollectionProduct $productCollection
+     */
+    private $productCollection;
```
Comment (rogyar) [inline app/code/Magento/CatalogUrlRewrite/Test/Integration/Model/Product/Attribute/Backend/IsUniqueUrlKeyTest.php]: Please, rename it to productCollectionFactory

File: app/code/Magento/CatalogUrlRewrite/Test/Integration/Model/Product/Attribute/Backend/IsUniqueUrlKeyTest.php [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+namespace Magento\Catalog\Test\Integration\Model\Product\Attribute\Backend;
+
+use Magento\Catalog\Api\ProductRepositoryInterface;
+use Magento\Store\Model\Store;
+use Magento\Store\Model\StoreManager;
+use Magento\TestFramework\Helper\Bootstrap;
+use Magento\Framework\Exception\LocalizedException;
+use PHPUnit\Framework\TestCase;
+use Magento\Catalog\Model\ResourceModel\Product as ProductResource;
+use Magento\Catalog\Model\ResourceModel\Product\CollectionFactory as CollectionProduct;
```
Comment (rogyar) [inline app/code/Magento/CatalogUrlRewrite/Test/Integration/Model/Product/Attribute/Backend/IsUniqueUrlKeyTest.php]: Please, name alias as ProductCollectionFactory

File: app/code/Magento/CatalogUrlRewrite/Setup/Patch/Data/UpdateUrlKeyProductAttributePatch.php [hunk 9]
```diff
@@ -0,0 +1,77 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
```
Comment (rogyar) [inline app/code/Magento/CatalogUrlRewrite/Setup/Patch/Data/UpdateUrlKeyProductAttributePatch.php]: Please, use strict types declaration for new implementations

File: app/code/Magento/CatalogUrlRewrite/Model/ResourceModel/Product/IsUniqueUrlKey.php [hunk 10]
```diff
@@ -0,0 +1,103 @@
+<?php
+
```
Comment (rogyar) [inline app/code/Magento/CatalogUrlRewrite/Model/ResourceModel/Product/IsUniqueUrlKey.php]: Please, add Magento copyright

File: app/code/Magento/CatalogUrlRewrite/Model/Product/Attribute/Backend/UrlKey.php [hunk 11]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+declare(strict_types=1);
```
Comment (rogyar) [inline app/code/Magento/CatalogUrlRewrite/Model/Product/Attribute/Backend/UrlKey.php:6]: Please, move the strict types declaration beyond the namespace declaration and add Magento copyright.

File: app/code/Magento/CatalogUrlRewrite/Model/ResourceModel/Product/IsUniqueUrlKey.php [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    /**
+     * @var ResourceConnection
+     */
+    private $resourceConnection;
+
+    /**
+     * @var \Magento\Eav\Model\Config
+     */
+    private $eavConfig;
```
Comment (rogyar) [inline app/code/Magento/CatalogUrlRewrite/Model/ResourceModel/Product/IsUniqueUrlKey.php:31]: Please, add extra empty line here

File: app/code/Magento/CatalogUrlRewrite/Model/ResourceModel/Product/IsUniqueUrlKey.php [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+                $this->resourceConnection->getTableName('catalog_product_entity_varchar')
+            )->where(
+                "attribute_id = $urlKeyAttributeId"
+            )->where(
+                "value = '$urlKeyValue'"
+            )->where(
+                "store_id = $storeId"
+            );
+        $result = $this->connection->fetchOne($qry);
+        if(is_bool($result)){
```
Comment (rogyar) [inline app/code/Magento/CatalogUrlRewrite/Model/ResourceModel/Product/IsUniqueUrlKey.php]: Please change the condition to match PSR-2 requirements (https://www.php-fig.org/psr/psr-2/#51-if-elseif-else)

File: app/code/Magento/CatalogUrlRewrite/Setup/Patch/Data/UpdateUrlKeyProductAttributePatch.php [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+namespace Magento\CatalogUrlRewrite\Setup\Patch\Data;
+
+use Magento\Catalog\Setup\CategorySetup;
+use Magento\Catalog\Setup\CategorySetupFactory;
+use Magento\Framework\Setup\ModuleDataSetupInterface;
+use Magento\Framework\Setup\Patch\DataPatchInterface;
+
+/**
+ * Class UpdateProductAttributes
+ * @package Magento\Catalog\Setup\Patch
```
Comment (rogyar) [inline app/code/Magento/CatalogUrlRewrite/Setup/Patch/Data/UpdateUrlKeyProductAttributePatch.php]: Please, remove the package annotation

File: app/code/Magento/CatalogUrlRewrite/Setup/Patch/Data/UpdateUrlKeyProductAttributePatch.php [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+        ModuleDataSetupInterface $moduleDataSetup,
+        CategorySetupFactory $categorySetupFactory
+    ) {
+        $this->moduleDataSetup = $moduleDataSetup;
+        $this->categorySetupFactory = $categorySetupFactory;
+    }
+
+    /**
+     * @inheritdoc
+     * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
```
Comment (rogyar) [inline app/code/Magento/CatalogUrlRewrite/Setup/Patch/Data/UpdateUrlKeyProductAttributePatch.php]: Please, remove this suppression annotation

File: app/code/Magento/CatalogUrlRewrite/Test/Integration/Model/Product/Attribute/Backend/IsUniqueUrlKeyTest.php [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+use Magento\Framework\Registry;
+
+/**
+ * Class IsUniqueUrlKeyTest
+ * @package Magento\Catalog\Test\Integration\Model\Product\Attribute\Backend
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
```
Comment (rogyar) [inline app/code/Magento/CatalogUrlRewrite/Test/Integration/Model/Product/Attribute/Backend/IsUniqueUrlKeyTest.php]: We don't need this annotation in our case

File: app/code/Magento/CatalogUrlRewrite/Test/Integration/Model/Product/Attribute/Backend/IsUniqueUrlKeyTest.php [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+use Magento\TestFramework\Helper\Bootstrap;
+use Magento\Framework\Exception\LocalizedException;
+use PHPUnit\Framework\TestCase;
+use Magento\Catalog\Model\ResourceModel\Product as ProductResource;
+use Magento\Catalog\Model\ResourceModel\Product\CollectionFactory as CollectionProduct;
+use Magento\Framework\Registry;
+
+/**
+ * Class IsUniqueUrlKeyTest
+ * @package Magento\Catalog\Test\Integration\Model\Product\Attribute\Backend
```
Comment (rogyar) [inline app/code/Magento/CatalogUrlRewrite/Test/Integration/Model/Product/Attribute/Backend/IsUniqueUrlKeyTest.php]: Please, remove the package annotation

**Review by rogyar — CHANGES_REQUESTED** [review]

**Comment by rogyar** [comment]
Hi @riccardougolini. Thank you for the clarification. Please, consider fixing minor code style issues mentioned in the review

**Comment by riccardougolini** [comment]
hi @rogyar, thank you for reviewing my solution. 
I applied your advice check them in the new patch.

**Pull request closed without merging**